diff -upr libhttpserver.clean/src/webserver.cpp libhttpserver/src/webserver.cpp
--- libhttpserver.clean/src/webserver.cpp	2023-06-16 03:49:56.000000000 +0000
+++ libhttpserver/src/webserver.cpp	2023-10-17 10:11:18.350231611 +0000
@@ -420,7 +420,7 @@ void* uri_log(void* cls, const char* uri
     std::ignore = con;
 
     auto mr = std::make_unique<details::modded_request>();
-    mr->complete_uri = std::make_unique<string>(uri);
+    mr->complete_uri = std::make_unique<string>(uri == NULL ? "" : uri);
     return reinterpret_cast<void*>(mr.release());
 }
 