diff -upr libhttpserver.clean/examples/Makefile.am libhttpserver/examples/Makefile.am
--- libhttpserver.clean/examples/Makefile.am	2023-06-16 03:49:56.000000000 +0000
+++ libhttpserver/examples/Makefile.am	2023-10-17 09:20:20.024117944 +0000
@@ -16,8 +16,12 @@
 #     License along with this library; if not, write to the Free Software
 #     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 
-LDADD = $(top_builddir)/src/libhttpserver.la
-AM_CPPFLAGS = -I$(top_srcdir)/src -I$(top_srcdir)/src/httpserver/
+DEPS_PATH=$(top_srcdir)/../../
+RE2_PATH=$(DEPS_PATH)/re2/re2
+RE2_IDIR=$(RE2_PATH)
+
+LDADD = $(top_builddir)/src/libhttpserver.la -L$(RE2_PATH)/obj -lre2
+AM_CPPFLAGS = -I$(top_srcdir)/src -I$(top_srcdir)/src/httpserver/ -I$(RE2_IDIR)
 METASOURCES = AUTO
 noinst_PROGRAMS = hello_world service minimal_hello_world custom_error allowing_disallowing_methods handlers hello_with_get_arg setting_headers custom_access_log basic_authentication digest_authentication minimal_https minimal_file_response minimal_deferred url_registration minimal_ip_ban benchmark_select benchmark_threads benchmark_nodelay deferred_with_accumulator file_upload
 
diff -upr libhttpserver.clean/src/details/http_endpoint.cpp libhttpserver/src/details/http_endpoint.cpp
--- libhttpserver.clean/src/details/http_endpoint.cpp	2023-06-16 03:49:56.000000000 +0000
+++ libhttpserver/src/details/http_endpoint.cpp	2023-10-17 09:36:36.625373755 +0000
@@ -43,6 +43,7 @@ http_endpoint::~http_endpoint() {
 }
 
 http_endpoint::http_endpoint(const string& url, bool family, bool registration, bool use_regex):
+    re_url_normalized(new re2::RE2("")),
     family_url(family),
     reg_compiled(false) {
     if (use_regex && !registration) {
@@ -110,8 +111,18 @@ http_endpoint::http_endpoint(const strin
     if (use_regex) {
         url_normalized += "$";
         try {
-            re_url_normalized = std::regex(url_normalized, std::regex::extended | std::regex::icase | std::regex::nosubs);
-        } catch (std::regex_error& e) {
+            re2::RE2::Options opts {};
+            opts.set_case_sensitive(true);
+            opts.set_never_capture(true);
+
+            re_url_normalized.reset(
+                new re2::RE2(url_normalized, opts)
+            );
+
+            if (re_url_normalized->ok() != true) {
+                throw std::invalid_argument("Invalid regex supplied to re2");
+            }
+        } catch (std::invalid_argument& e) {
             throw std::invalid_argument("Not a valid regex in URL: " + url_normalized);
         }
         reg_compiled = true;
@@ -124,7 +135,7 @@ http_endpoint::http_endpoint(const http_
     url_pars(h.url_pars),
     url_pieces(h.url_pieces),
     chunk_positions(h.chunk_positions),
-    re_url_normalized(h.re_url_normalized),
+    re_url_normalized(new re2::RE2(h.re_url_normalized->pattern())),
     family_url(h.family_url),
     reg_compiled(h.reg_compiled) {
 }
@@ -134,7 +145,7 @@ http_endpoint& http_endpoint::operator =
     url_normalized = h.url_normalized;
     family_url = h.family_url;
     reg_compiled = h.reg_compiled;
-    re_url_normalized = h.re_url_normalized;
+    re_url_normalized.reset(new re2::RE2(h.re_url_normalized->pattern()));
     url_pars = h.url_pars;
     url_pieces = h.url_pieces;
     chunk_positions = h.chunk_positions;
@@ -150,7 +161,7 @@ bool http_endpoint::match(const http_end
     if (!reg_compiled) throw std::invalid_argument("Cannot run match. Regex suppressed.");
 
     if (!family_url || url.url_pieces.size() < url_pieces.size()) {
-        return regex_match(url.url_complete, re_url_normalized);
+        return RE2::FullMatch(url.url_complete, *re_url_normalized);
     }
 
     string nn = "/";
@@ -159,7 +170,7 @@ bool http_endpoint::match(const http_end
         nn += (first ? "" : "/") + url.url_pieces[i];
         first = false;
     }
-    return regex_match(nn, re_url_normalized);
+    return RE2::FullMatch(nn, *re_url_normalized);
 }
 
 }  // namespace details

diff -upr libhttpserver.clean/src/httpserver/details/http_endpoint.hpp libhttpserver/src/httpserver/details/http_endpoint.hpp
--- libhttpserver.clean/src/httpserver/details/http_endpoint.hpp	2023-06-16 03:49:56.000000000 +0000
+++ libhttpserver/src/httpserver/details/http_endpoint.hpp	2023-10-17 09:40:50.077082067 +0000
@@ -27,11 +27,12 @@
 
 // cpplint errors on regex because it is replaced (in Chromium) by re2 google library.
 // We don't have that alternative here (and we are actively avoiding dependencies).
-#include <regex>  // NOLINT [build/c++11]
+#include "re2/re2.h"
 #include <stdexcept>
 #include <string>
 #include <utility>
 #include <vector>
+#include <memory>
 
 namespace httpserver {
 
@@ -128,7 +129,7 @@ class http_endpoint {
      http_endpoint():
          url_complete("/"),
          url_normalized("/"),
-         re_url_normalized(std::regex("")),  // initialize empty
+         re_url_normalized(new re2::RE2("")), // initialize empty
          family_url(false),
          reg_compiled(false) { }
 
@@ -177,7 +178,7 @@ class http_endpoint {
      /**
       * Regex used in comparisons
      **/
-     std::regex re_url_normalized;
+     std::unique_ptr<re2::RE2> re_url_normalized;
 
      /**
       * Boolean indicating wheter the endpoint represents a family
diff -upr libhttpserver.clean/src/Makefile.am libhttpserver/src/Makefile.am
--- libhttpserver.clean/src/Makefile.am	2023-06-16 03:49:56.000000000 +0000
+++ libhttpserver/src/Makefile.am	2023-10-17 09:20:20.024117944 +0000
@@ -16,7 +16,11 @@
 #     License along with this library; if not, write to the Free Software
 #     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 
-AM_CPPFLAGS = -I../ -I$(srcdir)/httpserver/
+DEPS_PATH=$(top_srcdir)/../../
+RE2_PATH=$(DEPS_PATH)/re2/re2
+RE2_IDIR=$(RE2_PATH)
+
+AM_CPPFLAGS = -I../ -I$(srcdir)/httpserver/ -I$(RE2_IDIR)
 METASOURCES = AUTO
 lib_LTLIBRARIES = libhttpserver.la
 libhttpserver_la_SOURCES = string_utilities.cpp webserver.cpp http_utils.cpp file_info.cpp http_request.cpp http_response.cpp string_response.cpp basic_auth_fail_response.cpp digest_auth_fail_response.cpp deferred_response.cpp file_response.cpp http_resource.cpp details/http_endpoint.cpp
@@ -37,7 +41,7 @@ endif
 
 libhttpserver_la_CFLAGS = $(AM_CFLAGS)
 libhttpserver_la_CXXFLAGS = $(AM_CXXFLAGS)
-libhttpserver_la_LDFLAGS = $(AM_LDFLAGS) -no-undefined
+libhttpserver_la_LDFLAGS = $(AM_LDFLAGS) -no-undefined -L$(RE2_PATH)/obj -lre2
 
 install-data-hook:
 	(mkdir -p $(DESTDIR)$(includedir) && cd $(DESTDIR)$(includedir) && $(LN_S) -f httpserver.hpp httpserverpp)
