diff --git include/ma_priv.h include/ma_priv.h
index 0fefd52e..6172d0ec 100644
--- include/ma_priv.h
+++ include/ma_priv.h
@@ -28,10 +28,9 @@ MYSQL_FIELD * unpack_fields(const MYSQL *mysql, MYSQL_DATA *data,
                             MA_MEM_ROOT *alloc,uint fields,
                             my_bool default_value);
 
-static inline my_bool ma_has_extended_type_info(const MYSQL *mysql)
+static inline my_bool ma_has_extended_type_info(const MYSQL *mysql __attribute__((unused)))
 {
-  return ((mysql->extension->mariadb_server_capabilities) &
-          (MARIADB_CLIENT_EXTENDED_METADATA >> 32)) != 0;
+  return 0;
 }
 
 static inline uint ma_extended_type_info_rows(const MYSQL *mysql)
