diff --git libmariadb/mariadb_lib.c libmariadb/mariadb_lib.c
index c9aec200..be4c91d1 100644
--- libmariadb/mariadb_lib.c
+++ libmariadb/mariadb_lib.c
@@ -1752,6 +1752,11 @@ restart:
     if (ma_pvio_write(pvio, (unsigned char *)hdr, len) <= 0)
     {
       ma_pvio_close(pvio);
+      if (mysql->options.extension->async_context &&
+        mysql->options.extension->async_context->pvio) {
+        /* async_context->pvio contains dangling pointer. Invalidatng pvio pointer in the async context*/
+        mysql->options.extension->async_context->pvio = NULL;
+      }
       goto error;
     }
   }
