--- /tmp/charset.c	2017-12-28 14:58:14.174281738 +0100
+++ unittest/libmariadb/charset.c	2017-12-28 14:59:53.510120540 +0100
@@ -74,12 +74,12 @@
   char *csname= (char*) "utf8";
   char *csdefault= (char*)mysql_character_set_name(mysql);
 
-  FAIL_IF(mysql_set_character_set(mysql, csname), mysql_error(mysql));
+  FAIL_IF(mysql_set_character_set(mysql, csname, 0), mysql_error(mysql));
 
   mysql_get_character_set_info(mysql, &cs);
 
   FAIL_IF(strcmp(cs.csname, "utf8") || strcmp(cs.name, "utf8_general_ci"), "Character set != UTF8");
-  FAIL_IF(mysql_set_character_set(mysql, csdefault), mysql_error(mysql));
+  FAIL_IF(mysql_set_character_set(mysql, csdefault, 0), mysql_error(mysql));
 
   return OK;
 }
@@ -542,7 +542,7 @@
   }
   /* Retrieve character set information. */
 
-  mysql_set_character_set(mysql, "latin1");
+  mysql_set_character_set(mysql, "latin1", 0);
   bug30472_retrieve_charset_info(mysql,
                                  character_set_name_1,
                                  character_set_client_1,
@@ -551,7 +551,7 @@
 
   /* Switch client character set. */
 
-  FAIL_IF(mysql_set_character_set(mysql, "utf8"), "Setting cs to utf8 failed");
+  FAIL_IF(mysql_set_character_set(mysql, "utf8", 0), "Setting cs to utf8 failed");
 
   /* Retrieve character set information. */
 
@@ -581,7 +581,7 @@
   /* Call mysql_change_user() with the same username, password, database. */
 
   rc= mysql_change_user(mysql, username, password, (schema) ? schema : "test");
-  mysql_set_character_set(mysql, "latin1");
+  mysql_set_character_set(mysql, "latin1", 0);
   check_mysql_rc(rc, mysql);
 
   /* Retrieve character set information. */
@@ -647,7 +647,7 @@
     /* ignore ucs2 */
     if (strcmp(row[0], "ucs2") && strcmp(row[0], "utf16le") && strcmp(row[0], "utf8mb4") && 
         strcmp(row[0], "utf16") && strcmp(row[0], "utf32")) {
-      rc= mysql_set_character_set(mysql, row[0]);
+      rc= mysql_set_character_set(mysql, row[0], 0);
       check_mysql_rc(rc, mysql);
     }
   }
@@ -748,7 +748,7 @@
 
   if (strcmp(osname, "utf8"))
   {
-    rc= mysql_set_character_set(mysql, "utf8");
+    rc= mysql_set_character_set(mysql, "utf8", 0);
     check_mysql_rc(rc, mysql);
 
     csname2= mysql_character_set_name(mysql);
@@ -756,7 +756,7 @@
 
     FAIL_IF(!strcmp(csname2, csname1), "Wrong charset: expected utf8");
 
-    rc= mysql_set_character_set(mysql, "auto");
+    rc= mysql_set_character_set(mysql, "auto", 0);
     check_mysql_rc(rc, mysql);
 
     csname2= mysql_character_set_name(mysql);
