--- cpp-dotenv.old/src/dotenv.cpp	2023-06-12 10:09:44.621343972 +0000
+++ cpp-dotenv.new/src/dotenv.cpp	2020-08-09 16:19:56.000000000 +0000
@@ -13,6 +13,7 @@
 
 dotenv::dotenv& dotenv::dotenv::load_dotenv(const string& dotenv_path, const bool overwrite, const bool interpolate)
 {
+    loaded = false;
     ifstream env_file;
     env_file.open(dotenv_path);
 
@@ -21,6 +22,7 @@
         Parser parser;
         parser.parse(env_file, overwrite, interpolate);
         env_file.close();
+        loaded = true;
     }
 
     return *this;
